### PR TITLE
Include files sync up from RISCV-CTG

### PR DESCRIPTION
**Pull request by yungchinghsiao**
This pull request attempts to sync up arch_test.h and encoding.h from RISCV-CTG

### CODE REVIEW

**Review by allenjbaum — CHANGES_REQUESTED** [review]
Unless I'm missing some context, this will break every test that isn't a floating point test
SREG is the width of the X registers; it doesn't change just because FLEN changes

**Review by allenjbaum — CHANGES_REQUESTED** [review]
For RVTEST_SIGUPD_P64: it is clearly the case that this is for an RV32 only. Therefore, you don't need to use SREG.  
And, in one case you  have an explicit offset of 4, and the other you use REGWIDTH. 
And, this doesn't check for offset overflow.
And this completely replicates the RVTEST_SIGUPD_FID macro, so just alias it.
RVTEST_SIGUPD_PK64 doesn't check overflow either, 
And, again you have an explicit offset of 4, but update using REGWIDTH. Be consistant.
And: REGWIDTH+REGWIDTH+REGWIDTH should simply be 3*REGWIDTH (or just be 12 if this can only be used for RV32)

**Review by allenjbaum — COMMENTED** [review]
TEST_RD_OP completely replicates TEST_CMV_OP. Just alias it

**Review by allenjbaum — CHANGES_REQUESTED** [review]
Unless I'm missing some context, this will break every test that isn't a floating point test
LREG, SREG is the width of the X registers; it doesn't change just because FLEN changes.
Was this intended to (re)define FSREG, FLREG?

**Comment by allenjbaum** [comment]
I would have them in separate PRs (after the macros get fixed).
I have added my requests for changes.

File: riscv-test-suite/env/arch_test.h [hunk 1]
```diff
@@ -95,6 +95,15 @@
...
 
+#if XLEN==64
+  #if FLEN==32
+    #define SREG sw
+    #define LREG lW
+    #define REGWIDTH 4
+    #define MASK 0xFFFFFFFF
+  #endif
+#endif
+
```
[the remark this answers is not in the document]
Comment (yungchinghsiao) [reply riscv-test-suite/env/arch_test.h]: I've removed these lines in the new changes in this PR.

**Comment by yungchinghsiao** [comment]
> For RVTEST_SIGUPD_P64: it is clearly the case that this is for an RV32 only. Therefore, you don't need to use SREG. And, in one case you have an explicit offset of 4, and the other you use REGWIDTH. And, this doesn't check for offset overflow. And this completely replicates the RVTEST_SIGUPD_FID macro, so just alias it. RVTEST_SIGUPD_PK64 doesn't check overflow either, And, again you have an explicit offset of 4, but update using REGWIDTH. Be consistant. And: REGWIDTH+REGWIDTH+REGWIDTH should simply be 3*REGWIDTH (or just be 12 if this can only be used for RV32)

Fixed by making these macros constants. and thanks for this review, I found that TEST_PKRR_OP() and TEST_PKR_OP() should have updated their flagreg to signatures.

**Comment by yungchinghsiao** [comment]
> TEST_RD_OP completely replicates TEST_CMV_OP. Just alias it

Done

**Comment by yungchinghsiao** [comment]
Sorry, accidentally closed this pull request. reopening it now.

**Comment by yungchinghsiao** [comment]
@allenjbaum @pawks: I wonder if you had any chance to look at my changes to address your review comments? Please do not hesitate to let me know if there are anything that I should do further.

**Comment by pawks** [comment]
I think all the old comments have been addressed and this PR can be merged.

**Review by allenjbaum — CHANGES_REQUESTED** [review]
These all update offset, but never check that it will cause an overflow.
I have this code at the beginning every SIGUPD macro:
  .if offset+{update_amt} >=2048   ;\
     addi   _BR, _BR,offset                 ;\
     .set   offset, 0				 ;\
  .endif						 ;\

These macros are much longer than they need to be because of the way you check for NARG(__VA_ARGS__) == 0,1. This becomes much more understandable if you instead use this;
  .if NARG(__VA_ARGS__) == 1				       ;\
     .set offset, _ARG1(__VA_OPT__(__VA_ARGS__,0))	;\
  .endif 
and then use offset after that point ( instead of ARG1() )while removing the other code conditional  on __ARG1 (well, remove one, and remove the .if/.endif from the other.)
I think I made a mistake asking for explicit offset and sw instead of SREG if you knew it was for RV32.  I was mostly trying to get consistent use. It's much safer to use REGWIDTH and SREG, as it enables these macros to be reused for the other length, but this code is not broken, so perhaps just leave that alone for now.

As I mentioned, RVTEST_SIGUPD_P64 is redundant and an alias of RVTEST_SIGUPD_FID, can be replaced by:
#define RVTEST_SIGUPD_P64(_BR,_R,_R_HI,...)		 \
 .if NARG(__VA_ARGS__) == 0				;\
	RVTEST_SIGUPD_FID(_BR,_R,_R_HI)			;\
 .else							;\
 	RVTEST_SIGUPD_FID(_BR,_R,_R_HI,_ARG1(__VA_OPT__(__VA_ARGS__,0)));\
 .endif

**Comment by yungchinghsiao** [comment]
Hello Allen,

I don't fully understand your comment on SIGUPD macros. Are you suggesting
me to do the following updates? If you could confirm this is what you mean,
I could help apply them to all SIGUPD macros (BTW, I thought offset
overflow was handled in generator.py in RISCV-CTG)

*diff --git a/riscv_ctg/env/arch_test.h b/riscv_ctg/env/arch_test.h*

*index a05966b..322de13 100644*

*--- a/riscv_ctg/env/arch_test.h*

*+++ b/riscv_ctg/env/arch_test.h*

@@ -573,14 +573,15 @@ rvtest_data_end:

 #define _ARG1(_1ST,...) _1ST

 #define NARG(...) _ARG5(__VA_OPT__(__VA_ARGS__,)4,3,2,1,0)

 #define RVTEST_SIGUPD(_BR,_R,...)\

+  .if offset+REGWIDTH >=2048;\

+    addi _BR, _BR, offset;\

+    .set offset, 0;\

+  .endif;\

   .if NARG(__VA_ARGS__) == 1;\

-    SREG _R,_ARG1(__VA_ARGS__,0)(_BR);\

-    .set offset,_ARG1(__VA_OPT__(__VA_ARGS__,)0)+REGWIDTH;\

+    .set offset,_ARG1(__VA_OPT__(__VA_ARGS__,)0);\

   .endif;\

-  .if NARG(__VA_ARGS__) == 0;\

-    SREG _R,offset(_BR);\

+  SREG _R,offset(_BR);\

   .set offset,offset+REGWIDTH;\

-  .endif;



 #define RVTEST_SIGUPD_F(_BR,_R,_F,...)\

   .if NARG(__VA_ARGS__) == 1;\





Or the modified macro:



#define RVTEST_SIGUPD(_BR,_R,...)\

  .if offset+REGWIDTH >=2048;\

    addi _BR, _BR, offset;\

    .set offset, 0;\

  .endif;\

  .if NARG(__VA_ARGS__) == 1;\

    .set offset,_ARG1(__VA_OPT__(__VA_ARGS__,)0);\

  .endif;\

  SREG _R,offset(_BR);\

  .set offset,offset+REGWIDTH;\

**Comment by allenjbaum** [comment]
CTG can keep track of it, but no one is required to use CTG, so we need to
make the macros bulletproof.
The code you listed for SIGUPD is what I mean. I modified all the SIGUPD
macros I could find in the file I sent you,
so you should be able to see them all, not just SIGUPD but also SIGUPD_F,
SIGUPD_FID, SIG_UPD_P64, etc.
You should just cut and paste that entire section with SIGUPD*,
[VAL]BASEUPD*, **WITH** all the comments.

**Comment by yungchinghsiao** [comment]
@allenjbaum I rebased the PR to the latest master and updated all of the  SIGUPD* macros, including the comments. Please check again.

**Comment by allenjbaum** [comment]
looks good

**Comment by allenjbaum** [comment]
This addresses all my issues

**Review by allenjbaum — APPROVED** [review]
This addresses all my issues

**Review by allenjbaum — APPROVED** [review]
Thisis exactly what was requested

**Pull request merged**